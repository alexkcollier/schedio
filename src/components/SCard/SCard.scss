.card {
  background-color: color('white');
  border-radius: border-radius();
  box-shadow: shadow('border'), shadow();
  display: flex;
  flex-direction: column;

  > :first-child {
    border-top-left-radius: inherit;
    border-top-right-radius: inherit;
  }

  > :last-child {
    border-bottom-left-radius: inherit;
    border-bottom-right-radius: inherit;
  }

  &-header {
    border-bottom: border();
    border-color: color('grey', 'light');
    flex: 1 1 auto;
    padding: spacing('double');
  }

  &-image {
    flex: 1 1 auto;
    width: 100%;
  }

  &-body {
    flex: 1 1 auto;
    padding: spacing('double');

    &--no-padding {
      padding: spacing('none');
    }
  }

  &-footer {
    border-top: border();
    border-color: color('grey', 'light');
    flex: 1 1 auto;
    padding: spacing('double');
  }
}
