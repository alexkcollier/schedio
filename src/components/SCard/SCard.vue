<template>
  <component :is="tag" class="card">
    <div v-if="$slots.header" class="card__header">
      <slot name="header"/>
    </div>

    <img
      v-if="image && image.src"
      v-bind="image"
      class="card__image">

    <div v-if="$slots.default" :class="{ 'card__body--no-padding': hasNoPadding }" class="card__body">
      <slot/>
    </div>

    <div v-if="$slots.footer" class="card__footer">
      <slot name="footer"/>
    </div>
  </component>
</template>

<script>
const imageValidator = obj => {
  const requiredAttrs = ['src', 'alt']

  return requiredAttrs.every(attr => {
    const hasAttr = obj.hasOwnProperty(attr)

    if (!hasAttr) console.error(`\`image\` requires attribute: ${attr}`)

    return hasAttr
  })
}

export default {
  name: 'SCard',

  props: {
    tag: {
      type: String,
      default: 'div'
    },

    image: {
      type: Object,
      default: () => ({ src: '', alt: '' }),
      validator: imageValidator
    },

    hasNoPadding: {
      type: Boolean,
      default: false
    }
  }
}
</script>
