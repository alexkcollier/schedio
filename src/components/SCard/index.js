import SCard from './SCard.vue';

export { SCard };

export default { install: Vue => Vue.component(SCard.name, SCard) };
