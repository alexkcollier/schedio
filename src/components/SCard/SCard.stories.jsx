import { SHeading } from '@/components/SHeading'
import { SCard } from '@/components/SCard'
import PropList from '@@/docs/components/PropList'
import { boolean, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/Layout.SCard', module)
  .addParameters({ jest: 'SCard' })
  .add('Card', () => {
    return {
      props: {
        image: {
          default: () => ({
            src: text('image.src', '', 'Optional Props'),
            alt: text('image.alt', '', 'Optional Props'),
          }),
        },
        hasNoPadding: { default: boolean('has-no-padding', false, 'Optional Props') },
        tag: { default: text('tag', 'div', 'Optional Props') },
        headerSlot: { default: text('Header', 'I am the header slot', 'Slots') },
        defaultSlot: { default: text('Default', 'I am the default slot', 'Slots') },
        footerSlot: { default: text('Footer', 'I am the footer slot', 'Slots') },
      },
      render (h) {
        const { headerSlot, defaultSlot, footerSlot, ...props } = this.$props

        return (
          <div>
            <SHeading level="1">Card</SHeading>
            <p>
              It's a card! It has header, footer, and default slots, and supports images. Padding on
              the default slot can be toggled with the <code>has-no-padding</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>
            <SCard {...{ props }}>
              <template slot="header">{headerSlot}</template>
              <template slot="default">{defaultSlot}</template>
              <template slot="footer">{footerSlot}</template>
            </SCard>

            <PropList component={SCard} />
          </div>
        )
      },
    }
  })
  .add(
    'Card Types',
    () => ({
      render (h) {
        return (
          <div>
            <SHeading>Card Types</SHeading>

            <SHeading level="2">Standard</SHeading>
            <p>All slots are optional.</p>
            <SCard>
              <template slot="header">Header</template>
              <template slot="default">Default</template>
              <template slot="footer">Footer</template>
            </SCard>

            <SHeading level="2">Image</SHeading>
            <p>You should only need to use the default or header slots with an image.</p>
            <SCard image={{ src: ' https://picsum.photos/800', alt: 'A random image' }}>
              <template slot="header">Header</template>
              <template slot="default">Default</template>
            </SCard>
          </div>
        )
      },
    }),
    {
      options: {
        showPanel: false,
      },
    }
  )
