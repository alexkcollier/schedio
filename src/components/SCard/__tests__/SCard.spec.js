import { shallowMount } from '@vue/test-utils'
import SCard from '@/components/SCard/SCard.vue'
import { axe, toHaveNoViolations } from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('SCard.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SCard, {
      slots: {
        default: '<div>Default Slot</div>',
      },
    })
  })

  it('renders correctly', async () => {
    expect(wrapper.contains('.card')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('can use different HTML tags', () => {
    wrapper.setProps({ tag: 'section' })
    expect(wrapper.contains('section')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can show an image', async () => {
    const props = {
      image: {
        src: '/some-image.jpg',
        alt: 'An image',
      },
    }

    wrapper.setProps(props)
    expect(wrapper.props('image')).toEqual(props.image)
    expect(wrapper.contains('img')).toBe(true)

    const imageAttributes = wrapper.find('img').attributes()

    expect(imageAttributes).toMatchObject(props.image)
    expect(wrapper.html()).toMatchSnapshot()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('can have content', () => {
    expect(wrapper.vm.$slots.default).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can remove padding from content', () => {
    wrapper.setProps({ hasNoPadding: true })
    expect(wrapper.contains('.card__body--no-padding')).toBe(true)
  })

  it('can have a header', () => {
    const withHeader = shallowMount(SCard, {
      slots: {
        header: '<div>Header Slot</div>',
      },
    })

    expect(withHeader.vm.$slots.header).toBeTruthy()
    expect(withHeader.html()).toMatchSnapshot()
  })

  it('can have a footer', () => {
    const withFooter = shallowMount(SCard, {
      slots: {
        footer: '<div>Footer Slot</div>',
      },
    })

    expect(withFooter.vm.$slots.footer).toBeTruthy()
    expect(withFooter.html()).toMatchSnapshot()
  })
})
