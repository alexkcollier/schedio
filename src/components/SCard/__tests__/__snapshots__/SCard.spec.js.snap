// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCard.vue can have a footer 1`] = `
<div class="card">
  <!---->
  <!---->
  <!---->
  <div class="card__footer">
    <div>Footer Slot</div>
  </div>
</div>
`;

exports[`SCard.vue can have a header 1`] = `
<div class="card">
  <div class="card__header">
    <div>Header Slot</div>
  </div>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`SCard.vue can have content 1`] = `
<div class="card">
  <!---->
  <!---->
  <div class="card__body">
    <div>Default Slot</div>
  </div>
  <!---->
</div>
`;

exports[`SCard.vue can show an image 1`] = `
<div class="card">
  <!----> <img alt="An image" src="/some-image.jpg" class="card__image">
  <div class="card__body">
    <div>Default Slot</div>
  </div>
  <!---->
</div>
`;

exports[`SCard.vue can use different HTML tags 1`] = `
<section class="card">
  <!---->
  <!---->
  <div class="card__body">
    <div>Default Slot</div>
  </div>
  <!---->
</section>
`;

exports[`SCard.vue renders correctly 1`] = `
<div class="card">
  <!---->
  <!---->
  <div class="card__body">
    <div>Default Slot</div>
  </div>
  <!---->
</div>
`;
